tools/db/repair-brig-clients-table/src/Work.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-orphans -Wno-unused-imports #-}
+
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2022 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Work where
+
+import Cassandra
+import Data.Conduit
+import Data.Conduit.Internal (zipSources)
+import Data.Conduit.List qualified as C
+import Data.Id
+import Data.Misc
+import Data.Time.Clock
+import Imports
+import System.Logger (Logger)
+import System.Logger qualified as Log
+import UnliftIO.Async (pooledMapConcurrentlyN)
+import Wire.API.Team.Feature
+import Wire.API.User
+
+runCommand :: Bool -> Logger -> ClientState -> IO ()
+runCommand dryRun l brig = do
+  runConduit $
+    zipSources
+      (C.sourceList [(1 :: Int32) ..])
+      (transPipe (runClient brig) getClients)
+      .| C.mapM
+        ( \(i, rows) -> do
+            Log.info l (Log.field "number of clients processed: " (show (i * pageSize)))
+            pure rows
+        )
+      .| C.mapM_ (\rows -> runClient brig (mapM_ (filterReportRemove dryRun l) rows))
+
+pageSize :: Int32
+pageSize = 1000
+
+type ClientRow =
+  ( UserId, -- user
+    Text, -- client
+    Maybe (Cassandra.Set Int32), -- capabilities
+    Maybe Int32, -- class
+    Maybe Text, -- cookie
+    Maybe Text, -- label
+    Maybe UTCTime, -- last_active
+    Maybe Text, -- model
+    Maybe UTCTime, -- tstamp
+    Maybe Int32 -- type
+  )
+
+getClients :: ConduitM () [ClientRow] Client ()
+getClients = paginateC cql (paramsP LocalQuorum () pageSize) x5
+  where
+    cql :: PrepQuery R () ClientRow
+    cql = "select user, client, capabilities, class, cookie, label, last_active, model, tstamp, type from clients"
```

Comment:
I'm wondering if *select all clients, then filter in Haskell* is the best strategy. Can't we limit the workload by a smarter query? I.e. one that queries for `Rows` with erroneous values?

Reply:
Okay, I'm used to using databases ... not these *"things"* :wink: 

As it's a one-time action, I think we don't have to spend more time on this :smile_cat: